boards: add Zolertia Firefly board

AnonMall opened the pull request:
I added the firefly board, with the remote-revb board as a template. I am still using remote-common, as the Firefly also has the Zoul core module installed. The gpio defines are adjusted according to #7320, which has to be merged first, because the spi and uart pins are depending on the changes, otherwise the board will not boot.

haukepetersen commented on boards/firefly/Makefile.include:
```diff
@@ -0,0 +1,11 @@
+include $(RIOTBOARD)/remote-common/Makefile.include
+
+# Board provides 2 network interfaces and should be 
+#GNRC_NETIF_NUMOF := 2
```
I think it would make sense to move this into the board's `Makefile.dep` and set the value according to the network devices that are actually enabled.

AnonMall replied:
Just a quick question about that. Is there a possibilty to choose which network device to enable at runtime? For example, I have finished the CC1200 network driver for the Zoul Module and the Firefly is able to use both interfaces at the same time, so booting up both and using both interfaces is possible when leaving `GNRC_NETIF_NUMOF := 2` in. But when i say `GNRC_NETIF_NUMOF := 1`, it always enables the CC1200 network device, if I leave the `USEMODULE += CC1200` in. So what I mean is, if it is possible to have both network devices, meaning the `CC1200` and the `CC2538 RF` enabled, while just using one interface by using `GNRC_NETIF_NUMOF := 1` and switch the corresponding device at runtime?

AnonMall replied:
But maybe its best to leave it out for now and add it later when the CC1200 driver has been merged.

haukepetersen commented on boards/firefly/Makefile.include:
```diff
@@ -0,0 +1,11 @@
+
+# Board provides 2 network interfaces and should be 
+#GNRC_NETIF_NUMOF := 2
+# after the CC1200 driver has been implemented
```
newline at EOF missing

haukepetersen commented on boards/firefly/board.c:
```diff
@@ -0,0 +1,51 @@
+    /* Shoot rainbows */
+    LED_RAINBOW();
+
+}
```
newline...

haukepetersen commented on boards/firefly/include/board.h:
```diff
@@ -0,0 +1,97 @@
+ * @{
+ */
+//#define SHUTDOWN_EN_GPIO    GPIO_PD1
+/** @} */
```
just remove this block. We can re-add it if it happens that this pin is actually used for something...

haukepetersen commented on boards/firefly/include/board.h:
```diff
@@ -0,0 +1,97 @@
+#define CC1200_RESET_GPIO   GPIO_PIN(2, 7)
+#define CC1200_GPD0_GPIO    GPIO_PIN(1, 4)
+#define CC1200_GPD2_GPIO    GPIO_PIN(1, 0)
+
```
`/** @} */` missing

haukepetersen commented on boards/firefly/include/periph_conf.h:
```diff
@@ -0,0 +1,116 @@
+        .miso_pin = GPIO_PIN(1, 3),
+        .sck_pin  = GPIO_PIN(1, 2),
+        //Commented out in order to omit hardware control for the CC1200
+        //.cs_pin   = GPIO_PB5
```
The `cs_pin` should be set in any case, so I would remove the two lines above and put `.cs_pin    = GPIO_UNEF` here

aabadie commented on boards/firefly/README.md:
```diff
@@ -0,0 +1,111 @@
+Zolertia Firefly platform
```
Very nice documentation but IMHO it would be better to put it either at the beginning of board.h or in a separate `doc.txt` file.

aabadie commented on boards/firefly/board.c:
```diff
@@ -0,0 +1,51 @@
+
+    /* Shoot rainbows */
+    LED_RAINBOW();
+
```
This empty line can be removed

aabadie commented on boards/firefly/include/periph_conf.h:
```diff
@@ -0,0 +1,127 @@
+/* UART 0 pin configuration */
+#define UART_0_TX_PIN       GPIO_PIN(0, 1)
+#define UART_0_RX_PIN       GPIO_PIN(0, 0)
+
```
This empty line can be removed

smlng commented on boards/firefly/include/periph_conf.h:
```diff
@@ -0,0 +1,127 @@
+/** @} */
+
+/**
+ * @name ADC configuration
```
remove, as it is empty or at least unconfigured?

AnonMall commented:
okay I removed the lines and corrected the style marks.
@aabadie I don't know about the documentation though. It is a fixed version of the `RE-Mote Rev.B` Readme. For the `board.h` it could be too long, but I wouldn't mind moving it to a `doc.txt`

aabadie commented on boards/firefly/board.c:
```diff
@@ -0,0 +1,51 @@
+void board_init(void)
+{
+    /* initialize the boards LEDs */
+    leds_init();
```
To be consistent with other boards, the gpio init functions should be called after the `cpu_init`

aabadie commented on boards/firefly/board.c:
```diff
@@ -0,0 +1,51 @@
+ */
+
+/**
+ * @defgroup    boards_firefly
```
copy-paste, defgroup should be used once in board.h, here only use `@ingroup  boards_firefly`

aabadie commented on boards/firefly/include/board.h:
```diff
@@ -0,0 +1,88 @@
+ * @name LED pin definitions
+ * @{
+ */
+#define LED0_PIN        GPIO_PIN(3, 5)
```
Is there an enum for the port ? something like PORT_A/B/C/D ?
It would be great to add a define for the port of each LED/BTN
```c
#define LED0_PORT PORT_D
```
for example

AnonMall replied:
the cc2538 gpio code is currently in change, the port enums are flagged as deprecated. See #7517

AnonMall replied:
Ah I actually just found the enums in the CC2538. I missed that. I will change it no problem

aabadie commented on boards/firefly/Makefile.include:
```diff
@@ -0,0 +1,7 @@
+
+# define the default port depending on the host OS
+PORT_LINUX  ?= /dev/ttyUSB0
+PORT_DARWIN ?= $(firstword $(sort $(wildcard /dev/tty.SLAB_USBtoUART*)))
```
Are you sure this is the correct port for mac OS ? can someone with a mac can confirm ?
on remote-pa (cc2538), it's `/dev/tty.usbserial`, on remote-reva/b it's also `/dev/tty.SLAB_USBtoUART`

AnonMall replied:
I am on Linux so would also be happy for someone on MacOS to check.

smlng replied:
typically that name depends on the USB2Serial converted used. According to the [data sheets](https://github.com/Zolertia/Resources/wiki/Zolertia-Technical-documentation#the-re-mote-platform-documentation) both (remote and firefly) use the CP2104, so I would assume names are also the same. Though I can't verify as I have no Firefly (only remotes) available 😞 

TL;DR leave as is, we'll fix if needed.

aabadie commented on boards/firefly/Makefile:
```diff
@@ -0,0 +1,5 @@
+MODULE = board
+
+DIRS = $(RIOTBOARD)/remote-common
```
Event if it's not the case in other places, I'm wondering if using `DIRS +=` wouldn't be better ?

AnonMall replied:
I agree, shouldn't do any harm. I will change it.

aabadie commented on boards/firefly/Makefile.features:
```diff
@@ -0,0 +1,14 @@
+FEATURES_PROVIDED += periph_timer
+FEATURES_PROVIDED += periph_uart
+
+# Various other features (if any)
```
Not needed as it's already provided by the CPU

aabadie commented on boards/firefly/Makefile.features:
```diff
@@ -0,0 +1,14 @@
+FEATURES_PROVIDED += cpp
+
+# The board MPU family (used for grouping by the CI system)
+FEATURES_MCU_GROUP = cortex_m3_2
```
Add the following at the end of this file to include CPU features:
```
-include $(RIOTCPU)/cc2538/Makefile.features
```

aabadie commented on boards/firefly/include/board.h:
```diff
@@ -0,0 +1,88 @@
+ * @{
+ */
+#define LED0_PIN        GPIO_PIN(3, 5)
+#define LED1_PIN        GPIO_PIN(3, 4)
```
If port macros exist, please replace `GPIO_PIN(<port idx>, X)` by `GPIO_PIN(<port macro>, X)` everywhere needed.

aabadie commented on boards/firefly/Makefile:
```diff
@@ -0,0 +1,5 @@
+MODULE = board
+
+DIRS += $(RIOTBOARD)/remote-common
```
A large refactoring of the board configuration was recently merged. Please update this file accordingly

aabadie commented on boards/firefly/Makefile.dep:
```diff
@@ -0,0 +1 @@
+include $(RIOTBOARD)/remote-common/Makefile.dep
```
Now thi should be include `$(RIOTBOARD)/common/remote/Makefile.dep`

smlng commented:
needs rebase and some work, remove milestone for now.

aabadie commented:
It would be great to have this soon. There are some firefly available on IoT-LAB in the Lille site.

AnonMall commented:
I will tend to it in the next week and hopefully have it finished by the end of the week as well. Sorry for the long delay.

AnonMall commented:
Actually I just rebased everything and worked the requested fixes into it. 
One important thing is, that with the new `GPIO_PIN()` definitions, the I2C code does not work anymore, as the CC2538 I2C code accesses the registers directly and is not compatible with the `gpio_t` representation.

tcschmidt commented:
@AnonMall do you consider updating?

AnonMall commented:
@tcschmidt As far as I know, the whole I2C API gets an overhaul, so it is better to wait for that to finish

kYc0o commented:
@AnonMall the i2c rework is finished. Do you mind to rebase your PR?

kYc0o commented:
I just tested it and it lacks the adc_params.h file to work with saul. You just need to copy this file from the remote-revb into the firefly/include folder.

Otherwise I think this PR might be merged.

kYc0o commented:
@AnonMall ping. If you don't have time or don't want to continue with this PR I propose myself to take over it. Thanks for your contribution!

AnonMall commented:
@kYc0o Sorry, for the late reply. I added the file you requested and performed a rebase.

kYc0o commented:
@aabadie and @haukepetersen can you take a look if your concerns were addressed?

kYc0o commented:
@AnonMall Everything looks ok for me now, just the commit message should be: `boards: add firefly`. Just for info you could add some lines below the commit message to indicate that this is part of the remote cc2538 based boards.

aabadie commented:
I could test this board with a one line command on IoT-LAB (thanks again @cladmi !):
```
make BOARD=firefly IOTLAB_NODE=firefly-1.lille.iot-lab.info -C examples/gnrc_networking flash term
```

I tested `examples/default` and `examples/gnrc_networking` with success. With `gnrc_networking`, I could ping 2 nodes and send/receive UDP packets. With `default`, I could read the ADC values and toggle the LEDs but I cannot confirm that the LEDs were actually toggling (because I can't see them ;) ).

Note that I also rebased locally your branch and had to apply the following patch to make it build (it also contains some cleanup):
```diff
diff --git a/boards/firefly/Makefile.features b/boards/firefly/Makefile.features
index bdfcd7e5b..24693fb34 100644
--- a/boards/firefly/Makefile.features
+++ b/boards/firefly/Makefile.features
@@ -1,10 +1,10 @@
 # Put defined MCU peripherals here (in alphabetical order)
+FEATURES_PROVIDED += periph_adc
 FEATURES_PROVIDED += periph_gpio
 FEATURES_PROVIDED += periph_i2c
 FEATURES_PROVIDED += periph_spi
 FEATURES_PROVIDED += periph_timer
 FEATURES_PROVIDED += periph_uart
-FEATURES_PROVIDED += periph_adc
 
 # The board MPU family (used for grouping by the CI system)
 FEATURES_MCU_GROUP = cortex_m3_2
diff --git a/boards/firefly/include/board.h b/boards/firefly/include/board.h
index 1e76cc25d..f9e951181 100644
--- a/boards/firefly/include/board.h
+++ b/boards/firefly/include/board.h
@@ -23,7 +23,6 @@
 
 #include "cpu.h"
 #include "board_common.h"
-#include "periph_cpu.h"
 
 #ifdef __cplusplus
 extern "C" {
@@ -34,21 +33,19 @@ extern "C" {
  * @{
  */
 #define LED0_PIN        GPIO_PIN(PORT_D, 5)
-#define LED1_PIN        GPIO_PIN(PORT_D, 4)
-#define LED2_PIN        GPIO_PIN(PORT_D, 3)
-
 #define LED0_MASK       (1 << 5)
-#define LED1_MASK       (1 << 4)
-#define LED2_MASK       (1 << 3)
-
 #define LED0_ON         (GPIO_D->DATA |=  LED0_MASK)
 #define LED0_OFF        (GPIO_D->DATA &= ~LED0_MASK)
 #define LED0_TOGGLE     (GPIO_D->DATA ^=  LED0_MASK)
 
+#define LED1_PIN        GPIO_PIN(PORT_D, 4)
+#define LED1_MASK       (1 << 4)
 #define LED1_ON         (GPIO_D->DATA |=  LED1_MASK)
 #define LED1_OFF        (GPIO_D->DATA &= ~LED1_MASK)
 #define LED1_TOGGLE     (GPIO_D->DATA ^=  LED1_MASK)
 
+#define LED2_PIN        GPIO_PIN(PORT_D, 3)
+#define LED2_MASK       (1 << 3)
 #define LED2_ON         (GPIO_D->DATA |=  LED2_MASK)
 #define LED2_OFF        (GPIO_D->DATA &= ~LED2_MASK)
 #define LED2_TOGGLE     (GPIO_D->DATA ^=  LED2_MASK)
@@ -58,7 +55,7 @@ extern "C" {
  * @name User button pin definition
  * @{
  */
-#define BTN0_PIN        GPIO_PIN(0, 3)
+#define BTN0_PIN        GPIO_PIN(PORT_A, 3)
 #define BTN0_MODE       GPIO_IN_PU
 /** @} */
 
@@ -84,5 +81,6 @@ void board_init(void);
 #ifdef __cplusplus
 } /* end extern "C" */
 #endif
+
 #endif /* BOARD_H */
 /** @} */
diff --git a/boards/firefly/include/periph_conf.h b/boards/firefly/include/periph_conf.h
index d7104e813..9799594c2 100644
--- a/boards/firefly/include/periph_conf.h
+++ b/boards/firefly/include/periph_conf.h
@@ -19,8 +19,6 @@
 #ifndef PERIPH_CONF_H
 #define PERIPH_CONF_H
 
-#include "cc2538_gpio.h"
-#include "periph_cpu.h"
 #include "periph_common.h"
 
 #ifdef __cplusplus
@@ -50,35 +48,20 @@ static const i2c_conf_t i2c_config[I2C_NUMOF] = {
 };
 /** @} */
 
-/**
- * @brief   Pre-calculated clock divider values based on a CLOCK_CORECLOCK (32MHz)
- *
- * Calculated with (CPSR * (SCR + 1)) = (CLOCK_CORECLOCK / bus_freq), where
- * 1 < CPSR < 255 and
- * 0 < SCR  < 256
- */
-static const spi_clk_conf_t spi_clk_config[] = {
-    { .cpsr = 10, .scr = 31 },  /* 100khz */
-    { .cpsr =  2, .scr = 39 },  /* 400khz */
-    { .cpsr =  2, .scr = 15 },  /* 1MHz */
-    { .cpsr =  2, .scr =  2 },  /* ~4.5MHz */
-    { .cpsr =  2, .scr =  1 }   /* ~10.7MHz */
-};
-
 /**
  * @name SPI configuration
  * @{
  */
 static const spi_conf_t spi_config[] = {
     {
-        .dev      = SSI0,
+        .num      = 0,
         .mosi_pin = GPIO_PIN(PORT_B, 1),
         .miso_pin = GPIO_PIN(PORT_B, 3),
         .sck_pin  = GPIO_PIN(PORT_B, 2),
         .cs_pin   = GPIO_UNDEF
     },
     {
-        .dev      = SSI1,
+        .num      = 1,
         .mosi_pin = GPIO_PIN(PORT_C, 5),
         .miso_pin = GPIO_PIN(PORT_C, 6),
         .sck_pin  = GPIO_PIN(PORT_C, 4),
@@ -93,7 +76,7 @@ static const spi_conf_t spi_config[] = {
  * @name ADC configuration
  * @{
  */
-#define SOC_ADC_ADCCON_REF  SOC_ADC_ADCCON_REF_AVDD5
+#define SOC_ADC_ADCCON3_EREF  SOC_ADC_ADCCON3_EREF_AVDD5
 
 static const adc_conf_t adc_config[] = {
     GPIO_PIN(PORT_A, 5),    /**< GPIO_PA5 = ADC1_PIN */
```

aabadie commented:
I pushed another couple of commits to fix Murdock build issues. @kYc0o, if you agree, may I can squash and go ?

cladmi commented:
@aabadie as this is a new board that you have with IoT-LAB, can you try running all the current tests with https://github.com/RIOT-OS/Release-Specs/pull/68 ?

kYc0o commented:
If the tests need to be run using the `make test` target, the board won't succeed most of tests due to the lack of a functional `make reset` along with `make term`.

cladmi commented:
@kYc0o indeed you are right.
On IoT-LAB it is also done through the BSL using the same port as the serial. So cannot be flashed/reset when the serial port is open.

aabadie commented:
> So cannot be flashed/reset when the serial port is open.

I confirm it doesn't work on IoT-LAB because of the BSL. But regarding this PR how should we proceed then ?

kYc0o commented:
I'd say it's ready, I just need to do a local test and we're all set.

aabadie commented:
> I'd say it's ready, I just need to do a local test and we're all set.

Great, let me know when your local test is done, so I can squash.

kYc0o commented:
I just tested it and everything is OK. However it needs a rebase.

kYc0o commented:
@haukepetersen ping. @aabadie can you ACK or dismiss your review?

aabadie commented:
ping @haukepetersen, we just need your ACK to go here.

aabadie commented:
@haukepetersen, last call before dismiss. You have 24h left before I merge.

aabadie commented:
With no reply from @haukepetersen after 3 calls, I dismissed his review.

Let's merge this PR now.
